hack/tools.mk

```diff
@@ -0,0 +1,109 @@
+TOOLS_BIN_DIR              := $(TOOLS_DIR)/bin
+CONTROLLER_GEN             := $(TOOLS_BIN_DIR)/controller-gen
+GEN_CRD_API_REFERENCE_DOCS := $(TOOLS_BIN_DIR)/gen-crd-api-reference-docs
+GOIMPORTS                  := $(TOOLS_BIN_DIR)/goimports
+GOLANGCI_LINT              := $(TOOLS_BIN_DIR)/golangci-lint
+GO_TO_PROTOBUF             := $(TOOLS_BIN_DIR)/go-to-protobuf
+HELM                       := $(TOOLS_BIN_DIR)/helm
+MOCKGEN                    := $(TOOLS_BIN_DIR)/mockgen
+OPENAPI_GEN                := $(TOOLS_BIN_DIR)/openapi-gen
+PROMTOOL                   := $(TOOLS_BIN_DIR)/promtool
+PROTOC_GEN_GOGO            := $(TOOLS_BIN_DIR)/protoc-gen-gogo
+SETUP_ENVTEST              := $(TOOLS_BIN_DIR)/setup-envtest
+YAML2JSON                  := $(TOOLS_BIN_DIR)/yaml2json
+YQ                         := $(TOOLS_BIN_DIR)/yq
+
+export TOOLS_BIN_DIR := $(TOOLS_BIN_DIR)
+export PATH := $(abspath $(TOOLS_BIN_DIR)):$(PATH)
+
+#########################################
+# Common                                #
+#########################################
+
+# Tool targets should declare go.mod as a prerequisite, if the tool's version is managed via go modules. This causes
+# make to rebuild the tool in the desired version, when go.mod is changed.
+# For tools where the version is not managed via go.mod, we use a file per tool and version as an indicator for make
+# whether we need to install the tool or a different version of the tool (make doesn't rerun the rule if the rule is
+# changed).
+
+# Use this "function" to add the version file as a prerequisite for the tool target: e.g.
+#   $(HELM): $(call tool_version_file,$(HELM),$(HELM_VERSION))
+tool_version_file = $(TOOLS_BIN_DIR)/.version_$(subst $(TOOLS_BIN_DIR)/,,$(1))_$(2)
+
+# This target cleans up any previous version files for the given tool and creates the given version file.
+# This way, we can generically determine, which version was installed without calling each and every binary explicitly.
+$(TOOLS_BIN_DIR)/.version_%:
+	@version_file=$@; rm -f $${version_file%_*}*
+	@touch $@
+
+.PHONY: clean-tools-bin
+clean-tools-bin:
+	rm -rf $(TOOLS_BIN_DIR)/*
+
+#########################################
+# Tools                                 #
+#########################################
+
+$(CONTROLLER_GEN): go.mod
+	go build -o $(CONTROLLER_GEN) sigs.k8s.io/controller-tools/cmd/controller-gen
+
+$(GEN_CRD_API_REFERENCE_DOCS): go.mod
+	go build -o $(GEN_CRD_API_REFERENCE_DOCS) github.com/ahmetb/gen-crd-api-reference-docs
+
+$(GOIMPORTS): go.mod
+	go build -o $(GOIMPORTS) golang.org/x/tools/cmd/goimports
+
+GOLANGCI_LINT_VERSION := v1.42.1
+$(GOLANGCI_LINT): $(call tool_version_file,$(GOLANGCI_LINT),$(GOLANGCI_LINT_VERSION))
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(TOOLS_BIN_DIR) $(GOLANGCI_LINT_VERSION)
+
+$(GO_TO_PROTOBUF): go.mod
+	go build -o $(GO_TO_PROTOBUF) k8s.io/code-generator/cmd/go-to-protobuf
+
+HELM_VERSION := v3.5.4
```

Comment:
Thanks, that's a good suggestion. It will also allow other repos that reuse `hack/tools.mk` to use a different version from what g/g uses.
However, this might also make things complicated for e.g. `generate-controllerregistration` or similar stuff where `helm` is used.

I'm a bit torn here. I pushed a new commit, that defines default versions in `hack/tools.mk` which can be overwritten in the top level Makefile. Please tell me what you think about this approach :)

Reply:
Hm, wasn't the intention of @stoyanr to just put the versions to the top? I like the fact that you can now overwrite the versions. But is re-declaring them in `Makefile` really necessary? Can't we just go with the default versions in this repo?

Reply:
> Hm, wasn't the intention of @stoyanr to just put the versions to the top? I like the fact that you can now overwrite the versions. But is re-declaring them in `Makefile` really necessary? Can't we just go with the default versions in this repo?

Ooops, I misread the comment 🙈 
I amended the last commit: kept the variables "overridable" but moved them to the top and removed the override in the top-level make file.